### PR TITLE
use BaseModel uniformly and Ser&Deser

### PR DESCRIPTION
**Pull request by better629**
Rebase and Modified version of `https://github.com/geekan/MetaGPT/tree/serialization`

Feat
1. update memory and so on to BaseModel
2. add builtin_class_name to role/action to support list of roles/actions due to inherited problem.
3. dump the serialized project into a structured folder to make it clear. btw, a whole serialized project file is also fine, but maybe huge if there are many messages.
4. add more unittests to check the value of recovered module.

TODO
1. There are some non-software roles and actions need to updated.
2. It will serialize by default. If need a trigger please comment. 

Doc
https://github.com/better629/MetaGPT-docs/blob/main/src/zh/guide/tutorials/breakpoint_recovery.md

ser&deser example project
[team.zip](https://github.com/geekan/MetaGPT/files/13726574/team.zip)

image

### CODE REVIEW

File: tests/metagpt/serialize_deserialize/test_memory.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                   instruct_content=ic_obj(**out_data),
+                   content="system design content",
+                   cause_by=WriteDesign)
+
+    memory = Memory()
+    memory.add_batch([msg1, msg2])
+
+    stg_path = serdeser_path.joinpath("team", "environment")
+    memory.serialize(stg_path)
+    assert stg_path.joinpath("memory.json").exists()
```
Comment (geekan) [inline tests/metagpt/serialize_deserialize/test_memory.py:60]: 这是一个约定吗？

File: metagpt/team.py [hunk 2]
```diff
@@ -33,6 +36,36 @@ class Team(BaseModel):
...
+        stg_path = SERDESER_PATH.joinpath("team") if stg_path is None else stg_path
+
+        team_info_path = stg_path.joinpath("team_info.json")
+        write_json_file(team_info_path, self.dict(exclude={"env": True}))
+
+        self.env.serialize(stg_path.joinpath("environment"))  # save environment alone
+
+    @classmethod
+    def recover(cls, stg_path: Path) -> "Team":
+        return cls.deserialize(stg_path)
```
Comment (geekan) [inline metagpt/team.py:49]: 建议去除这个method，避免额外的理解

File: metagpt/utils/serialize.py [hunk 3]
```diff
@@ -35,14 +33,45 @@ def actionoutout_schema_to_mapping(schema: Dict) -> Dict:
...
+    new_mapping = {}
+    for key, value in mapping.items():
+        if value == "(<class 'str'>, Ellipsis)":
+            new_mapping[key] = (str, ...)
+        else:
+            new_mapping[key] = eval(value)  # `"'(list[str], Ellipsis)"` to `(list[str], ...)`
+    return new_mapping
+
+
+def serialize_general_message(message: "Message") -> dict:
```
Comment (geekan) [inline metagpt/utils/serialize.py:60]: is independent function for ser/deser message necessary?

File: metagpt/roles/role.py [hunk 4]
```diff
@@ -74,37 +80,20 @@ def values(cls):
...
+    # # env exclude=True to avoid `RecursionError: maximum recursion depth exceeded in comparison`
+    env: "Environment" = Field(default=None, exclude=True)
+    # TODO judge if ser&deser
+    msg_buffer: MessageQueue = Field(default_factory=MessageQueue,
+                                     exclude=True)  # Message Buffer with Asynchronous Updates
     memory: Memory = Field(default_factory=Memory)
     # long_term_memory: LongTermMemory = Field(default_factory=LongTermMemory)
     state: int = Field(default=-1)  # -1 indicates initial or termination state where todo is None
-    todo: Action = Field(default=None)
+    todo: Action = Field(default=None, exclude=True)
```
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/roles/role.py:93]: Yes. we can use `state` to remember next-todo one.

File: metagpt/roles/qa_engineer.py [hunk 5]
```diff
@@ -30,21 +37,21 @@
...
-        constraints="The test code you write should conform to code standard like PEP8, be modular, easy to read and maintain",
-        test_round_allowed=5,
-    ):
-        super().__init__(name, profile, goal, constraints)
-        self._init_actions(
-            [WriteTest]
-        )  # FIXME: a bit hack here, only init one action to circumvent _think() logic, will overwrite _think() in future updates
+    name: str = Field(default="Edward")
+    profile: str = Field(default="QaEngineer")
+    goal: str = "Write comprehensive and robust tests to ensure codes will work as expected without bugs"
```
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/roles/qa_engineer.py:42]: will use the Field uniformly

File: tests/metagpt/serialize_deserialize/test_memory.py (hunk 1 above)
Comment (better629) [reply tests/metagpt/serialize_deserialize/test_memory.py:60]: yes, see the doc's serialize data structure

File: metagpt/actions/write_prd.py [hunk 6]
```diff
@@ -64,15 +65,16 @@
...
+    content: Optional[str] = None
+    llm: BaseGPTAPI = Field(default_factory=LLM)
 
     async def run(self, with_messages, schema=CONFIG.prompt_schema, *args, **kwargs) -> ActionOutput | Message:
         # Determine which requirement documents need to be rewritten: Use LLM to assess whether new requirements are
         # related to the PRD. If they are related, rewrite the PRD.
         docs_file_repo = CONFIG.git_repo.new_file_repository(relative_path=DOCS_FILE_REPO)
         requirement_doc = await docs_file_repo.get(filename=REQUIREMENT_FILENAME)
-        if await self._is_bugfix(requirement_doc.content):
+        if requirement_doc and await self._is_bugfix(requirement_doc.content):
```
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/actions/write_prd.py:77]: very close, in old version, there is no restriction to add context in WriteCode. If write_code = WriteCode, `run` will failed.

File: metagpt/roles/engineer.py [hunk 7]
```diff
@@ -66,25 +69,21 @@ class Engineer(Role):
...
     """
+    name: str = "Alex"
+    profile: str = Field(default="Engineer")
+    goal: str = "write elegant, readable, extensible, efficient code"
+    constraints: str = "the code should conform to standards like google-style and be modular and maintainable. " \
+                       "Use same language as user requirement"
+    n_borg: int = 1
+    use_code_review: bool = False
+    code_todos: list = []
+    summarize_todos = []
```
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/roles/engineer.py:80]: add Field or not rule here:
if the field need to define `default_factory` or `exclude`, we use Field. else not to use Field.

File: metagpt/memory/memory.py [hunk 8]
```diff
@@ -6,20 +6,46 @@
...
     """The most basic memory: super-memory"""
-
-    def __init__(self):
-        """Initialize an empty storage list and an empty index dictionary"""
-        self.storage: list[Message] = []
-        self.index: dict[str, list[Message]] = defaultdict(list)
+    storage: list[Message] = Field(default=[])
+    index: dict[str, list[Message]] = Field(default_factory=defaultdict(list))
+
+    def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/memory/memory.py:25]: we serialize the object as json object and re-init or deserialize it through `__init__`.

pydantic (v1.x, we currently use) doc: https://docs.pydantic.dev/1.10/usage/exporting_models/
pydantic (v2.x) doc: https://docs.pydantic.dev/latest/concepts/serialization/  use `model_dump`.

reason to use init
1. like below example
2. init a Role and define _watch under `__init__`.

>>> from pydantic import BaseModel
>>> class A(BaseModel):
...     name: str = "wang"
...     age: int = 12
...
>>> a = A()
>>> a
A(name='wang', age=12)
>>> a.dict()
{'name': 'wang', 'age': 12}
>>> b = A(**{'name': 'tttang', 'age': 22})
>>> b
A(name='tttang', age=22)
>>> b = A("222", 44)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 332, in pydantic.main.BaseModel.__init__
TypeError: __init__() takes exactly 1 positional argument (3 given)
>>> b = A(name="222", age=44)

File: metagpt/utils/serialize.py (hunk 3 above)
Comment (better629) [reply metagpt/utils/serialize.py:60]: Message is a little different, in old version, dynamic BaseModel class like `prd` created by `create_model_class` in `instruct_content` can't be directly deserialized.
There two function will be rm due to not used in main code.

**Comment by geekan** [comment]
LGTM

**Pull request merged**